### ISSUE TITLE
linking problem with PNG library

### ISSUE THREAD

**Issue by amwink** [davisking/dlib#1736]
I am using the PNG writing function of Dlib (cloned from github into `/usr/local/dlib` and `dtest` run successfully for all tests but 2 machine precision errors) and tried to write a test image in the same way as the functions in `dilb/test`:

    #define DLIB_PNG_SUPPORT

    #include <dlib/image_io.h>                         
    #include <dlib/array2d.h>

    using namespace dlib;

    int main() {
  
      array2d<unsigned char>     img;
  
      img.set_size             ( 100,        100 );
      assign_all_pixels        ( img,          7 );
      save_png                 ( img, "test.png" );
   
      return (0);

    }
The command line I used for compiling this test program is 

    g++ -o test_dlib_png test_dlib_png.cpp /usr/local/dlib/dlib/all/source.cpp -I/usr/local/dlib -L/usr/lib/x86_64-linux-gnu -lpng16 -lpthread -lX11 
So I added the compilation of `/usr/local/dlib/dlib/all/source.cpp` and included `/usr/local/dlib` for the headers and `/usr/lib/x86_64-linux-gnu` for libraries, and linked against `libpng16'.

I expected to get a small greyscale image like `test.png` in the `dlib/test/build` directory. But at the moment I cannot compile the code with the `g++` call above, because it gives a linker error:

    /usr/bin/ld: /tmp/ccrXNLK3.o: in function `dlib::disable_if<dlib::is_matrix<dlib::array2d<unsigned char, dlib::memory_manager_stateless_kernel_1<char> >, void>, void>::type dlib::save_png<dlib::array2d<unsigned char, dlib::memory_manager_stateless_kernel_1<char> > >(dlib::array2d<unsigned char, dlib::memory_manager_stateless_kernel_1<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
    test_dlib_png.cpp:(.text._ZN4dlib8save_pngINS_7array2dIhNS_33memory_manager_stateless_kernel_1IcEEEEEENS_10disable_ifINS_9is_matrixIT_vEEvE4typeERKS7_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE[_ZN4dlib8save_pngINS_7array2dIhNS_33memory_manager_stateless_kernel_1IcEEEEEENS_10disable_ifINS_9is_matrixIT_vEEvE4typeERKS7_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE]+0x2a9): undefined reference to `dlib::impl::impl_save_png(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<unsigned char*, std::allocator<unsigned char*> >&, long, dlib::impl::png_type, int)'
    collect2: error: ld returned 1 exit status
In other words. If I have a C++ file as above called test_dlib_png.cpp in /tmp and try to compile it with the command above, I get a linker error.

Does it need to be linked against something other than libpng (or libpng16)?

Dlib version is >= 19.15, cloned this month from `https://github.com/davisking/dlib`
My platform is Debian 9.8, 64 bit with C++ compiler g++ 8.2.0-20

**Comment by davisking** [comment]
Never #define DLIB_PNG_SUPPORT inside a cpp file.  You have set it in one translation unit and not in others, creating ODR violations.  https://en.cppreference.com/w/cpp/language/definition

You have to set it everywhere, you typically do this via a compiler build switch or by letting a build system do it, since a build system will compile all translation units the same way, and therefore avoiding ODR violations.  This is why I encourage people to use cmake.  CMake won't create ODR violations be invoking the compiler in ways that violate ODR.  Or at least you will have to work to make it mess up the build like that.
